Only retrieve relevant volumes

jamiehannaford opened the pull request:
**What this PR does / why we need it**:

Improves performance for Cinder volume attach/detach calls. 

Currently when Cinder volumes are attached or detached, functions try to retrieve details about the volume from the Nova API. Because some only have the volume name not its UUID, they use the list function in gophercloud to iterate over all volumes to find a match. This incurs severe performance problems on OpenStack projects with lots of volumes (sometimes thousands) since it needs to send a new request when the current page does not contain a match. A better way of doing this is use the `?name=XXX` query parameter to refine the results.

**Which issue this PR fixes**:

https://github.com/kubernetes/kubernetes/issues/26404

**Special notes for your reviewer**:

There were 2 ways of addressing this problem:

1. Use the `name` query parameter
2. Instead of using the list function, switch to using volume UUIDs and use the GET function instead. You'd need to change the signature of a few functions though, such as [`DeleteVolume`](https://github.com/kubernetes/kubernetes/blob/master/pkg/volume/cinder/cinder.go#L49), so I'm not sure how backwards compatible that is.

Since #1 does effectively the same as #2, I went with it because it ensures BC.

One assumption that is made is that the `volumeName` being retrieved matches exactly the name of the volume in Cinder. I'm not sure how accurate that is, but I see no reason why cloud providers would want to append/prefix things arbitrarily. 

**Release note**:
```release-note
Improves performance of Cinder volume attach/detach operations
```

Commit:
Refactor to use Volume IDs and remove ambiguity

k8s-reviewable commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.kubernetes.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes/41785)
<!-- Reviewable:end -->

jingxu97 commented on pkg/cloudprovider/providers/openstack/openstack_volumes.go:
```diff
@@ -115,7 +115,8 @@ func (os *OpenStack) getVolume(diskName string) (volumes.Volume, error) {
 
-	err = volumes.List(sClient, nil).EachPage(func(page pagination.Page) (bool, error) {
+	opts := &volumes.ListOpts{Name: diskName}
+	err = volumes.List(sClient, opts).EachPage(func(page pagination.Page) (bool, error) {
```
What is the difference of using opts vs. nil?

jamiehannaford replied:
The Cinder API allows you to provide query parameters in the URL when retrieving a list of volumes. Filtering by [name](https://github.com/openstack/cinder/blob/master/cinder/api/v2/volumes.py#L109-L110) for example allows you to restrict the amount of volumes down considerably. Otherwise, the K8s code will traverse the entire collection of volumes in an OpenStack project to find a match (some project could have 10s of thousands). Filtering by name reduces/eliminates the need for subsequent pagination calls.

jamiehannaford commented:
Any status update for this?

jamiehannaford commented:
@k8s-sig-openstack-pr-reviews Would somebody mind reviewing this?

xsgordon commented:
@idvoretskyi can you add the sig/openstack label please?

hogepodge commented:
lgtm

jamiehannaford commented:
@hogepodge Would you mind re-running tests? There seems to be a flaky test case failing

jamiehannaford commented:
@jingxu97 @xsgordon @grodrigues3 @k8s-sig-openstack-api-reviews Would somebody mind commenting with `ok to test` so I can fix flaky tests?

jingxu97 commented:
/approve

jingxu97 commented:
@k8s-bot ok to test

jamiehannaford commented:
@jingxu97 @anguslees Would somebody mind commenting with `/approve` and `/lgtm`?

anguslees commented on pkg/cloudprovider/providers/openstack/openstack_volumes.go:
```diff
@@ -111,7 +111,9 @@ func (volumes *VolumesV2) createVolume(opts VolumeCreateOpts) (string, error) {
 	var volume Volume
-	err := volumes_v1.List(volumes.blockstorage, nil).EachPage(func(page pagination.Page) (bool, error) {
+
+	opts := &volumes_v1.ListOpts{Name: diskName}
```
The logic below looks for either name == diskName *or* the ID contains diskName.  That latter case sounds amazingly terrible, but are we confident we don't need it? (and if so, can we remove it in this PR to make it clear that that logic isn't going to trigger anymore?)

anguslees commented on pkg/cloudprovider/providers/openstack/openstack_volumes.go:
```diff
@@ -149,7 +151,9 @@ func (volumes *VolumesV1) getVolume(diskName string) (Volume, error) {
 	var volume Volume
-	err := volumes_v2.List(volumes.blockstorage, nil).EachPage(func(page pagination.Page) (bool, error) {
+
+	opts := &volumes_v2.ListOpts{Name: diskName}
```
Ditto - clarify the "ID contains diskName" case below

jamiehannaford commented:
@anguslees I've refactored a bunch of signatures to make the volume ID stuff clearer. I don't why it was implemented to handle both names and IDs, but I've switched it to the latter. It also seems that the current volume implementation uses IDs too, so I think this is backwards compatible.

anguslees reviewed (commented):
a) I think moving from names to IDs is absolutely the right thing to do here
b) I like that you've cleaned up all the symbols to make them more consistent (thanks!)

~I'm concerned about backwards compatibility across this migration~ - we still need to be able to attach a k8s volume created from older k8s.  I *think* this means that the few "public" mount/delete/etc (not provision) entry points need to be able to fall back to some form of get-id-by-name if lookup by id returns not-found.

Ideally that fallback lookup can be in a narrow set of clearly-commented entry points so that we know that by-name is a deprecated legacy (that we might, eg remove in release+1 with appropriate release notes).

.. And I'm sorry to ask you to put back some of the code you just worked hard to delete :(

Edit:
oh wait, you commented on exactly this above :)  ... Yes, I agree the existing behaviour uses IDs too (`openstack_volumes.go:createVolume()`) and it looks like it's been this way since ~forever (I traced it back as far as 2f06ebf9b76 - 2015).  So forget what I said above, I don't believe there's a back-compat issue here - until someone else raises some other volume-creation-path I don't know about.

/lgtm

jingxu97 commented:
/approve

jamiehannaford commented:
/assign @jsafrane @mikedanese

jsafrane commented:
/approve

jsafrane commented on pkg/cloudprovider/providers/rackspace/rackspace.go:
```diff
@@ -595,23 +571,23 @@ func (rs *Rackspace) DetachDisk(instanceID string, partialDiskId string) error {
+		glog.V(2).Infof("Successfully detached volume: %s from compute: %s", volume.ID, instanceID)
 	} else {
-		errMsg := fmt.Sprintf("Disk: %s has no attachments or is not attached to compute: %s", disk.Name, instanceID)
+		errMsg := fmt.Sprintf("Disk: %s has no attachments or is not attached to compute: %s", volume.Name, instanceID)
```
`volume.ID` is logged in all other cases, is there a reason for `volume.Name`?

jamiehannaford replied:
Probably not, but it was already there. This PR was just concerned with clarifying how UUIDs are used in function signatures.

jsafrane commented:
I checked the PR and if I understand it correctly, mostly it just renames variables and parameter names from `name` (`diskName`, ...) to `id` (`volumeID`, ...). The only behavior change is speeding up `getVolume()`, right? The PR is quite long and it renames too many things at once...

jamiehannaford commented:
@jsafrane This PR touches aspects of the current naming convention which is far from optimal. To quote another maintainer who reviewed this, the current logic is "amazingly terrible". Since the behaviour change interacted with this ID/name confusion, the PR seeks to also simplify it. I don't see a reason why this PR can't do both. 

From a wider perspective this patch has been approved by 2-3 other maintainers and despite that has been pending for over 2 months, so I'd really appreciate if we could merge this ASAP.

jamiehannaford commented:
@mikedanese Would you mind weighing in here and if necessary approving this change?

jsafrane commented:
> Since the behaviour change interacted with this ID/name confusion, the PR seeks to also simplify it. I don't see a reason why this PR can't do both.

I am not against the PR, I am trying to find out what it tries to do, as it is quite a long one and I do not want to miss some important spot hidden between renames.

So far it looks good, please rebase and I'll mark it as lgtm.

jamiehannaford commented:
@jsafrane Thanks for the feedback. I've made those changes - would you mind taking another quick look and adding `/lgtm`?

jsafrane commented:
/lgtm

jamiehannaford commented:
@mikedanese Can you `/approve` this PR?

mikedanese commented:
/approve

jamiehannaford commented:
@k8s-bot pull-kubernetes-federation-e2e-gce test this

jamiehannaford commented:
@k8s-bot pull-kubernetes-federation-e2e-gce test this

jamiehannaford commented:
@jsafrane Can you add `/lgtm` again? sorry for the pestering 😛

jamiehannaford commented:
Bump cc/ @jsafrane

jsafrane commented:
/lgtm
